### PR TITLE
Check approximate size before split check

### PR DESCRIPTION
**Pull request by huachaohuang**
Check approximate size before split check.

### CODE REVIEW

File: src/raftstore/store/util.rs [hunk 1]
```diff
@@ -89,6 +94,42 @@ pub fn is_epoch_stale(epoch: &metapb::RegionEpoch, check_epoch: &metapb::RegionE
...
+                                region: &metapb::Region)
+                                -> Result<TablePropertiesCollection> {
+    let cf = try!(get_cf_handle(db, cfname));
+    let start = keys::enc_start_key(region);
+    let end = keys::enc_end_key(region);
+    let range = Range::new(&start, &end);
+    db.get_properties_of_tables_in_range(cf, &[range]).map_err(|e| e.into())
+}
+
+pub fn get_region_approximate_size_cf(db: &DB,
```
Comment (siddontang) [inline src/raftstore/store/util.rs]: any test？

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1481,13 +1481,25 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            // TODO: Should we remove the `size_diff_hint`?
             if peer.size_diff_hint < self.cfg.region_check_size_diff {
                 continue;
             }
             info!("{} region's size diff {} >= {}, need to check whether should split",
                   peer.tag,
                   peer.size_diff_hint,
                   self.cfg.region_check_size_diff);
+
+            if let Ok(size) = peer.approximate_size() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: I prefer using another PR for split check

File: tests/raftstore/cluster.rs [hunk 3]
```diff
@@ -603,6 +603,12 @@ impl<T: Simulator> Cluster<T> {
         assert_eq!(resp.get_responses()[0].get_cmd_type(), CmdType::DeleteRange);
     }
 
+    pub fn must_flush(&mut self, sync: bool) {
+        for db in &self.dbs {
+            db.flush(sync).unwrap();
```
Comment (siddontang) [inline tests/raftstore/cluster.rs]: need flush_cf ?

File: src/raftstore/store/store.rs [hunk 4]
```diff
@@ -1520,10 +1520,20 @@ impl<T: Transport, C: PdClient> Store<T, C> {
             if peer.size_diff_hint < self.cfg.region_check_size_diff {
                 continue;
             }
-            info!("{} region's size diff {} >= {}, need to check whether should split",
-                  peer.tag,
-                  peer.size_diff_hint,
-                  self.cfg.region_check_size_diff);
+
+            match peer.approximate_size() {
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: do we need to check this in the split check worker so we don't need to block the raft store thread if calculating size may be slow?

File: tests/raftstore/cluster.rs (hunk 3 above)
Comment (huachaohuang) [inline tests/raftstore/cluster.rs]: I don't need to use `flush_cf`, if I add it, clippy will give me a warning.

File: src/raftstore/store/store.rs (hunk 4 above)
Comment (huachaohuang) [inline src/raftstore/store/store.rs]: OK

**Comment by huachaohuang** [comment]
PTAL @siddontang @zhangjinpeng1987

File: src/raftstore/store/worker/split_check.rs [hunk 5]
```diff
@@ -166,20 +159,39 @@ impl<C> Runner<C> {
...
 impl<C: Sender<Msg>> Runnable<Task> for Runner<C> {
     fn run(&mut self, task: Task) {
+        let region = &task.region;
+        let region_id = region.get_id();
+
+        // Check approximate size before scanning region.
+        match util::get_region_approximate_size(&self.engine, region) {
+            Ok(size) => {
+                if size < self.region_max_size {
+                    return;
```
Comment (siddontang) [inline src/raftstore/store/worker/split_check.rs]: need to add a metric here, like `CHECK_SPILT_COUNTER_VEC.with_label_values(&["skip"]).inc();`

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

**Comment by siddontang** [comment]
CI failed 

```
test raftstore::test_split_region::test_server_auto_split_region ... thread 'raftstore::test_split_region::test_server_auto_split_region' panicked at 'assertion failed: `(left != right)`
  left: `id: 1 start_key: "" end_key: "" region_epoch {conf_ver: 1 version: 1} peers {id: 5 store_id: 5} peers {id: 2 store_id: 2} peers {id: 3 store_id: 3} peers {id: 4 store_id: 4} peers {id: 1 store_id: 1}`,
 right: `id: 1 start_key: "" end_key: "" region_epoch {conf_ver: 1 version: 1} peers {id: 5 store_id: 5} peers {id: 2 store_id: 2} peers {id: 3 store_id: 3} peers {id: 4 store_id: 4} peers {id: 1 store_id: 1}`', tests/raftstore/test_split_region.rs:180:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:380
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:396
   4: tikv::util::panic_hook::track_hook::{{closure}}
             at src/util/panic_hook.rs:61
   5: <std::thread::local::LocalKey<T>>::try_with
             at /checkout/src/libstd/thread/local.rs:362
   6: <std::thread::local::LocalKey<T>>::with
             at /checkout/src/libstd/thread/local.rs:276
   7: tikv::util::panic_hook::track_hook
             at src/util/panic_hook.rs:55
   8: core::ops::function::Fn::call
             at /checkout/src/libcore/ops/function.rs:48
   9: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
  10: std::panicking::begin_panic
             at /checkout/src/libstd/panicking.rs:571
  11: std::panicking::begin_panic_fmt
             at /checkout/src/libstd/panicking.rs:521
  12: tests::raftstore::test_split_region::test_auto_split_region
             at tests/raftstore/test_split_region.rs:180
  13: tests::raftstore::test_split_region::test_server_auto_split_region
             at tests/raftstore/test_split_region.rs:227
  14: <F as test::FnBox<T>>::call_box
             at /checkout/src/libtest/lib.rs:1477
             at /checkout/src/libcore/ops/function.rs:143
             at /checkout/src/libtest/lib.rs:138
  15: __rust_maybe_catch_panic
             at /checkout/src/libpanic_unwind/lib.rs:98
```

I guess this may be caused by the change.

**Pull request merged**